Upgrade to React 19

avatus opened the pull request:
All `type-check` errors are fixed, and the app runs without issues. only about 6 tests left to fix. But getting this out there for review now to ensure we aren't fixing tests that may not need it/might break again if other changes are needed.

most of the changes are due to code mods and small manual changes

avatus committed:
update package.json

avatus committed:
codemod scoped-jsx

avatus committed:
refobject-defaults

gzdunek committed:
Fix some tests

avatus committed:
Lint

ravicious commented on package.json:
There's a bunch of unmet peer deps.

```
$ pnpm i --resolution-only
Scope: all 7 workspace projects
 WARN  `node_modules` is present. Lockfile only installation will make it out-of-date
 WARN  8 deprecated subdependencies found: @npmcli/move-file@2.0.1, abab@2.0.6, boolean@3.2.0, domexception@4.0.0, glob@7.2.3, glob@8.1.0, inflight@1.0.6, rimraf@3.0.2
Progress: resolved 1611, reused 1, downloaded 0, added 0, done
 ERR_PNPM_PEER_DEP_ISSUES  Unmet peer dependencies

.
├─┬ @nivo/bar 0.88.0
│ ├── ✕ unmet peer react@">= 16.14.0 < 19.0.0": found 19.1.0
│ ├─┬ @nivo/axes 0.88.0
│ │ ├── ✕ unmet peer react@">= 16.14.0 < 19.0.0": found 19.1.0
│ │ └─┬ @nivo/core 0.88.0
│ │   ├── ✕ unmet peer react@">= 16.14.0 < 19.0.0": found 19.1.0
│ │   └─┬ @nivo/tooltip 0.88.0
│ │     ├── ✕ unmet peer react@">= 16.14.0 < 19.0.0": found 19.1.0
│ │     └─┬ @nivo/core 0.88.0
│ │       ├── ✕ unmet peer react@">= 16.14.0 < 19.0.0": found 19.1.0
│ │       └─┬ @nivo/tooltip 0.88.0
│ │         ├── ✕ unmet peer react@">= 16.14.0 < 19.0.0": found 19.1.0
│ │         └─┬ @react-spring/web 9.7.5
│ │           ├── ✕ unmet peer react@"^16.8.0 || ^17.0.0 || ^18.0.0": found 19.1.0
│ │           ├── ✕ unmet peer react-dom@"^16.8.0 || ^17.0.0 || ^18.0.0": found 19.1.0
│ │           └─┬ @react-spring/core 9.7.5
│ │             ├── ✕ unmet peer react@"^16.8.0 || ^17.0.0 || ^18.0.0": found 19.1.0
│ │             ├─┬ @react-spring/shared 9.7.5
│ │             │ └── ✕ unmet peer react@"^16.8.0 || ^17.0.0 || ^18.0.0": found 19.1.0
│ │             └─┬ @react-spring/animated 9.7.5
│ │               └── ✕ unmet peer react@"^16.8.0 || ^17.0.0 || ^18.0.0": found 19.1.0
│ ├─┬ @nivo/legends 0.88.0
│ │ ├── ✕ unmet peer react@">= 16.14.0 < 19.0.0": found 19.1.0
│ │ └─┬ @nivo/colors 0.88.0
│ │   └── ✕ unmet peer react@">= 16.14.0 < 19.0.0": found 19.1.0
│ └─┬ @nivo/annotations 0.88.0
│   └── ✕ unmet peer react@">= 16.14.0 < 19.0.0": found 19.1.0
└─┬ react-day-picker 8.10.1
  └── ✕ unmet peer react@"^16.8.0 || ^17.0.0 || ^18.0.0": found 19.1.0
```

avatus replied:
nivo bar most recent release `0.93.0` still has unmet peer deps which makes me think this is some out of date library in general.

also, react-day-picker needs an update too.

avatus replied:
react-day-picker update here https://github.com/gravitational/teleport/pull/54632

this can merge before it tho (although preferably before 18 test plan)

ravicious commented on web/packages/design/src/Modal/Modal.tsx:
```diff
@@ -48,7 +48,7 @@ export type ModalProps = {
    * be adjusted if needed.
    */
-  children?: React.ReactElement;
+  children?: React.ReactElement<any>;
```
`React.ReactElement` seems to work just fine and in a number of other cases I checked.

https://react.dev/blog/2024/04/25/react-19-upgrade-guide#changes-to-the-reactelement-typescript-type

The way I understand it is that it's not needed in most cases. I think we should revert this. Unfortunately, this change is bundled with the codemod to `useRef`.

avatus replied:
i reverted the ones that didnt cause issues without .looks like mostly the `MenuAction` components require it

ravicious replied:
I reverted a few more of them, should be fine now!

ravicious commented on web/packages/shared/components/AccessRequests/NewRequest/RequestCheckout/RequestCheckout.tsx:
```diff
@@ -83,7 +83,7 @@ export const RequestCheckoutWithSlider = forwardRef<
     ref
   ) => {
-    const wrapperRef = useRef<HTMLDivElement>();
+    const wrapperRef = useRef<HTMLDivElement>(undefined);
```
IIRC, `null` should be used as the initial value for refs that hold DOM elements since that what React sets after a ref is removed from the DOM. https://react.dev/reference/react/useRef#manipulating-the-dom-with-a-ref

This codemod seems to be a bit overzealous. VSCode says that there's 43 instances of `useRef.*html.*undefined`, those should all probably be converted to `useRef<some type>(null)`. I should be able to do that with find and replace though, no?

I just don't want to leave dozens of `useRef<some HTML type>(undefined)` in the codebase. People are likely going to copy it but

1) It's not a good practice.
2) From what I see, the overload for `useRef(undefined)` was added [just to ease the migration](https://react.dev/blog/2024/04/25/react-19-upgrade-guide#useref-requires-argument).

avatus replied:
To be fair, some of it is codemod and some of it was manual work on me. I should have split them.

avatus replied:
Also without that undefined it was throwing errors but I'll double check

ravicious replied:
I replaced `useRef<HTML(.*)>\(undefined\)` with `useRef<HTML$1>(null)` in VSCode, it should be fine now!

ravicious committed:
Move import after license comment

ravicious committed:
Fix test types for a wrapper of Button

I couldn't get it to work when destructuring rest props, e.g.:

    const { size, children, ...buttonProps } = props;

so for now I just removed it altogether. I don't know why it works
in ForwardedAsWrapper, I suppose it has to do either with the css prop
passed to ButtonPrimary? Or maybe the forwardedAs prop accepted by
ForwardedAsWrapper?

avatus committed:
revert most ReactElement<any>

kimlisa committed:
Fix tests

kimlisa commented:
i managed to fix some tests, but i'm stuck on fixing the last remaining `bots/list/bots.test.tsx`. i'm stumped. we have a similar test called `roles.test.tsx`, and i tried using similar approach, but doesn't work... 🤔

kimlisa commented on web/packages/teleport/src/Account/Account.test.tsx:
```diff
@@ -370,11 +370,13 @@ test('removing an MFA method', async () => {
 
   await renderComponent(ctx);
-  await user.click(
+  fireEvent.click(
```
i don't understand why i had to use fireEvent here... (another file also required a fireEvent)

avatus replied:
thats funny you bring that up. @gzdunek had the opposite experience and managed to get `userEvent` to work better than `fireEvent`. YMMV i guess 🤷

gzdunek replied:
Yeah, in general it's better to use `userEvent` over `fireEvent`, since it's a higher-level API.
If `userEvent` isn't working, the issue must be somewhere else 🤔

kimlisa replied:
idk what happened in between me submitting the fix and reverting the fireEvents... reverting the fireEvent back to userEvent just works now.

i won't question it.

kimlisa replied:
ohhh!

ravicious commented on package.json:
> i managed to fix some tests, but i'm stuck on fixing the last remaining `bots/list/bots.test.tsx`. i'm stumped. we have a similar test called `roles.test.tsx`, and i tried using similar approach, but doesn't work... 🤔

I'm looking at this. It seems like that `MenuActionButton`'s onClick isn't called at all.

avatus replied:
ive been stuck on this all morning and its just not clicking no matter trying `userEvent` , `fireEvent`, and a native click `actionButton[0].click()`

avatus replied:
i also tried switching to something like `onMouseDown` but i havent found any documentation on why this would/wouldn't work after switching to 19 either way

ravicious replied:
Something super funky is going on here. I managed to trigger onClick of `MenuActionButton` by manually doing `element.click()`. However, even with that, the state of this component just never gets updated.

<details>
<summary>Patch</summary>

Copy then `pbpaste | git apply`.

```diff
diff --git a/web/packages/shared/components/MenuAction/MenuActionButton.tsx b/web/packages/shared/components/MenuAction/MenuActionButton.tsx
index ca679c16845..daa522a0d9b 100644
--- a/web/packages/shared/components/MenuAction/MenuActionButton.tsx
+++ b/web/packages/shared/components/MenuAction/MenuActionButton.tsx
@@ -51,15 +51,18 @@ export default class MenuActionIcon extends React.Component<
 
   onOpen = (e: React.MouseEvent<HTMLButtonElement>) => {
     e.stopPropagation();
+    console.log('opening MenuActionButton');
     this.setState({ open: true });
   };
 
   onClose = () => {
+    console.log('closing MenuActionButton');
     this.setState({ open: false });
   };
 
   render() {
     const { open } = this.state;
+    console.log('rendering MenuActionButton, open:', open);
     const { children, menuProps, buttonProps, icon } = this.props;
     return (
       <>
@@ -74,6 +77,7 @@ export default class MenuActionIcon extends React.Component<
           </FilledButtonIcon>
         ) : (
           <ButtonBorder
+            data-testid="menu-action-button"
             size="small"
             setRef={e => (this.anchorEl = e)}
             onClick={this.onOpen}
diff --git a/web/packages/teleport/src/Bots/List/Bots.test.tsx b/web/packages/teleport/src/Bots/List/Bots.test.tsx
index 2170d55bed6..b944c6b790b 100644
--- a/web/packages/teleport/src/Bots/List/Bots.test.tsx
+++ b/web/packages/teleport/src/Bots/List/Bots.test.tsx
@@ -16,10 +16,19 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+import { act } from 'react';
 import { MemoryRouter } from 'react-router';
 
-import { render, screen, userEvent, waitFor } from 'design/utils/testing';
+import {
+  prettyDOM,
+  render,
+  screen,
+  userEvent,
+  waitFor,
+  within,
+} from 'design/utils/testing';
 import { InfoGuidePanelProvider } from 'shared/components/SlidingSidePanel/InfoGuide';
+import { wait } from 'shared/utils/wait';
 
 import { botsApiResponseFixture } from 'teleport/Bots/fixtures';
 import { ContextProvider } from 'teleport/index';
@@ -81,29 +90,35 @@ test('shows missing permissions error if user lacks permissions to list', async
   ).toBeInTheDocument();
 });
 
-test('calls edit endpoint', async () => {
+test.only('calls edit endpoint', async () => {
+  const user = userEvent.setup();
   jest
     .spyOn(api, 'get')
-    .mockResolvedValueOnce({ ...botsApiResponseFixture })
+    .mockResolvedValueOnce({ items: [botsApiResponseFixture.items[0]] })
     .mockResolvedValueOnce(['role-1', 'editor']);
   jest.spyOn(api, 'put').mockResolvedValue({});
   renderWithContext(<Bots />);
 
-  await waitFor(() => {
-    expect(
-      screen.getByText(botsApiResponseFixture.items[0].metadata.name)
-    ).toBeInTheDocument();
+  const nameCell = await screen.findByText(
+    botsApiResponseFixture.items[0].metadata.name
+  );
+  const row = nameCell.closest('tr');
+  const button = within(row).getByText('Options');
+  console.log('Options button', prettyDOM(button));
+  await act(async () => {
+    button.click();
   });
+  // console.log('Options', prettyDOM(within(row).getByText('Options')));
+  // await user.click(within(row).getByText('Options'));
+  // console.log('actionCells[0]', prettyDOM(actionCells[0]));
+  // await user.click(actionCells[0]);
 
-  const actionCells = screen.queryAllByRole('button', { name: 'Options' });
-  expect(actionCells).toHaveLength(botsApiResponseFixture.items.length);
-  await userEvent.click(actionCells[0]);
-
-  expect(screen.getByText('Edit...')).toBeInTheDocument();
-  await userEvent.click(screen.getByText('Edit...'));
+  await wait(500);
+  console.log('all of DOM', prettyDOM(undefined, 10_000));
+  await user.click(await screen.findByText('Edit...'));
 
   expect(screen.getByText('Edit Bot')).toBeInTheDocument();
-  await userEvent.click(screen.queryByRole('button', { name: 'Save' }));
+  await user.click(screen.queryByRole('button', { name: 'Save' }));
 
   expect(screen.queryByText('Edit Bot')).not.toBeInTheDocument();
   expect(api.put).toHaveBeenCalledWith(
@@ -113,6 +128,7 @@ test('calls edit endpoint', async () => {
 });
 
 test('calls delete endpoint', async () => {
+  const user = userEvent.setup();
   jest
     .spyOn(api, 'get')
     .mockResolvedValueOnce({ ...botsApiResponseFixture })
@@ -128,15 +144,13 @@ test('calls delete endpoint', async () => {
 
   const actionCells = screen.queryAllByRole('button', { name: 'Options' });
   expect(actionCells).toHaveLength(botsApiResponseFixture.items.length);
-  await userEvent.click(actionCells[0]);
+  await user.click(actionCells[0]);
 
   expect(screen.getByText('Delete...')).toBeInTheDocument();
-  await userEvent.click(screen.getByText('Delete...'));
+  await user.click(screen.getByText('Delete...'));
 
   expect(screen.getByText('Delete Bot?')).toBeInTheDocument();
-  await userEvent.click(
-    screen.queryByRole('button', { name: 'Yes, Delete Bot' })
-  );
+  await user.click(screen.queryByRole('button', { name: 'Yes, Delete Bot' }));
 
   expect(screen.queryByText('Delete Bot?')).not.toBeInTheDocument();
   expect(api.delete).toHaveBeenCalledWith(

```

ravicious replied:
The issue with `<Bots>` is that it gets rendered twice which initializes `BotList` twice. So the click we make on the Options button essentially doesn't count, because the component is soon re-rendered from scratch. Maybe `BotList` is actually rendered, then unmounted and then rendered again because of some async shenanigans?

You cannot see it in the story, because the story uses just `BotList`, not `Bots`. But you can add a bot (`tctl bots add robot --roles=access --logins=foo`) and see it for yourself with my changes after opening the list of bots in the Web UI.

We either need to get rid of that unmounting (haven't checked if that's what happens, but I assume it's that) or simply wait until it happens in the test as a workaround?

Here's where my code is at:

<details>
<summary>Patch</summary>

Copy then `pbpaste | git apply`.

```diff
diff --git a/web/packages/shared/components/MenuAction/MenuActionButton.tsx b/web/packages/shared/components/MenuAction/MenuActionButton.tsx
index ca679c16845..2524edd0f55 100644
--- a/web/packages/shared/components/MenuAction/MenuActionButton.tsx
+++ b/web/packages/shared/components/MenuAction/MenuActionButton.tsx
@@ -45,21 +45,27 @@ export default class MenuActionIcon extends React.Component<
   };
 
   constructor(props: Props) {
+    console.log('initializing MenuActionButton');
     super(props);
-    this.state.open = props.defaultOpen || false;
+    this.state = { open: props.defaultOpen || false };
+    this.onOpen = this.onOpen.bind(this);
+    this.onClose = this.onClose.bind(this);
   }
 
-  onOpen = (e: React.MouseEvent<HTMLButtonElement>) => {
+  onOpen(e: React.MouseEvent<HTMLButtonElement>) {
     e.stopPropagation();
+    console.log('opening MenuActionButton');
     this.setState({ open: true });
-  };
+  }
 
-  onClose = () => {
+  onClose() {
+    console.log('closing MenuActionButton');
     this.setState({ open: false });
-  };
+  }
 
   render() {
     const { open } = this.state;
+    console.log('rendering MenuActionButton, open:', open);
     const { children, menuProps, buttonProps, icon } = this.props;
     return (
       <>
@@ -74,6 +80,7 @@ export default class MenuActionIcon extends React.Component<
           </FilledButtonIcon>
         ) : (
           <ButtonBorder
+            data-testid="menu-action-button"
             size="small"
             setRef={e => (this.anchorEl = e)}
             onClick={this.onOpen}
diff --git a/web/packages/teleport/src/Bots/List/BotList.tsx b/web/packages/teleport/src/Bots/List/BotList.tsx
index b3894d962ec..a4847f27e2d 100644
--- a/web/packages/teleport/src/Bots/List/BotList.tsx
+++ b/web/packages/teleport/src/Bots/List/BotList.tsx
@@ -70,26 +70,29 @@ export function BotList({
           },
           {
             altKey: 'options-btn',
-            render: bot => (
-              <BotOptionsCell
-                bot={bot}
-                onClickView={() => {
-                  setSelectedBot(bot);
-                  setInteraction(Interaction.VIEW);
-                }}
-                disabledEdit={disabledEdit}
-                disabledDelete={disabledDelete}
-                onClickEdit={() => {
-                  setSelectedBot(bot);
-                  setSelectedRoles(bot.roles);
-                  setInteraction(Interaction.EDIT);
-                }}
-                onClickDelete={() => {
-                  setSelectedBot(bot);
-                  setInteraction(Interaction.DELETE);
-                }}
-              />
-            ),
+            render: bot => {
+              console.log('rendering options button cell');
+              return (
+                <BotOptionsCell
+                  bot={bot}
+                  onClickView={() => {
+                    setSelectedBot(bot);
+                    setInteraction(Interaction.VIEW);
+                  }}
+                  disabledEdit={disabledEdit}
+                  disabledDelete={disabledDelete}
+                  onClickEdit={() => {
+                    setSelectedBot(bot);
+                    setSelectedRoles(bot.roles);
+                    setInteraction(Interaction.EDIT);
+                  }}
+                  onClickDelete={() => {
+                    setSelectedBot(bot);
+                    setInteraction(Interaction.DELETE);
+                  }}
+                />
+              );
+            },
           },
         ]}
         emptyText="No Bots Found"
diff --git a/web/packages/teleport/src/Bots/List/Bots.test.tsx b/web/packages/teleport/src/Bots/List/Bots.test.tsx
index 2170d55bed6..9df855898da 100644
--- a/web/packages/teleport/src/Bots/List/Bots.test.tsx
+++ b/web/packages/teleport/src/Bots/List/Bots.test.tsx
@@ -16,10 +16,19 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+import { act } from 'react';
 import { MemoryRouter } from 'react-router';
 
-import { render, screen, userEvent, waitFor } from 'design/utils/testing';
+import {
+  prettyDOM,
+  render,
+  screen,
+  userEvent,
+  waitFor,
+  within,
+} from 'design/utils/testing';
 import { InfoGuidePanelProvider } from 'shared/components/SlidingSidePanel/InfoGuide';
+import { wait } from 'shared/utils/wait';
 
 import { botsApiResponseFixture } from 'teleport/Bots/fixtures';
 import { ContextProvider } from 'teleport/index';
@@ -81,29 +90,37 @@ test('shows missing permissions error if user lacks permissions to list', async
   ).toBeInTheDocument();
 });
 
-test('calls edit endpoint', async () => {
+test.only('calls edit endpoint', async () => {
+  const user = userEvent.setup();
   jest
     .spyOn(api, 'get')
-    .mockResolvedValueOnce({ ...botsApiResponseFixture })
+    .mockResolvedValueOnce({ items: [botsApiResponseFixture.items[0]] })
     .mockResolvedValueOnce(['role-1', 'editor']);
   jest.spyOn(api, 'put').mockResolvedValue({});
   renderWithContext(<Bots />);
 
-  await waitFor(() => {
-    expect(
-      screen.getByText(botsApiResponseFixture.items[0].metadata.name)
-    ).toBeInTheDocument();
+  const nameCell = await screen.findByText(
+    botsApiResponseFixture.items[0].metadata.name
+  );
+  const row = nameCell.closest('tr');
+  const button = within(row).getByText('Options');
+  // console.log('Options button', prettyDOM(button));
+  await act(async () => {
+    button.click();
   });
+  // console.log('Options', prettyDOM(within(row).getByText('Options')));
+  // await user.click(within(row).getByText('Options'));
+  // console.log('actionCells[0]', prettyDOM(actionCells[0]));
+  // await user.click(actionCells[0]);
 
-  const actionCells = screen.queryAllByRole('button', { name: 'Options' });
-  expect(actionCells).toHaveLength(botsApiResponseFixture.items.length);
-  await userEvent.click(actionCells[0]);
-
-  expect(screen.getByText('Edit...')).toBeInTheDocument();
-  await userEvent.click(screen.getByText('Edit...'));
+  await act(async () => {
+    await wait(500);
+  });
+  // console.log('all of DOM', prettyDOM(undefined, 10_000));
+  await user.click(await screen.findByText('Edit...'));
 
   expect(screen.getByText('Edit Bot')).toBeInTheDocument();
-  await userEvent.click(screen.queryByRole('button', { name: 'Save' }));
+  await user.click(screen.queryByRole('button', { name: 'Save' }));
 
   expect(screen.queryByText('Edit Bot')).not.toBeInTheDocument();
   expect(api.put).toHaveBeenCalledWith(
@@ -113,6 +130,7 @@ test('calls edit endpoint', async () => {
 });
 
 test('calls delete endpoint', async () => {
+  const user = userEvent.setup();
   jest
     .spyOn(api, 'get')
     .mockResolvedValueOnce({ ...botsApiResponseFixture })
@@ -128,15 +146,13 @@ test('calls delete endpoint', async () => {
 
   const actionCells = screen.queryAllByRole('button', { name: 'Options' });
   expect(actionCells).toHaveLength(botsApiResponseFixture.items.length);
-  await userEvent.click(actionCells[0]);
+  await user.click(actionCells[0]);
 
   expect(screen.getByText('Delete...')).toBeInTheDocument();
-  await userEvent.click(screen.getByText('Delete...'));
+  await user.click(screen.getByText('Delete...'));
 
   expect(screen.getByText('Delete Bot?')).toBeInTheDocument();
-  await userEvent.click(
-    screen.queryByRole('button', { name: 'Yes, Delete Bot' })
-  );
+  await user.click(screen.queryByRole('button', { name: 'Yes, Delete Bot' }));
 
   expect(screen.queryByText('Delete Bot?')).not.toBeInTheDocument();
   expect(api.delete).toHaveBeenCalledWith(

```

ravicious replied:
K, it's actually `Table` that gets rendered twice and it initializes `MenuActionButton` twice. This is likely because that component is rendered through the `render` function in `Table`'s `columns` prop.

ravicious replied:
I removed the `useEffect` from `useTable` and the test now passes with no changes.

ravicious committed:
Remove extra useEffect from useTable

ravicious committed:
Pass null to useRef when dealing with DOM elements

ravicious committed:
Bring back just `ReactElement` without `<any>` where possible

ravicious commented on package.json:
VNet info tab doesn't look right, I'll check what's wrong in a sec.

Left is master, right is this branch.

<img width="2560" alt="vnet-marketing-tab" src="https://github.com/user-attachments/assets/23191c0f-a5f4-4c45-8dbb-263a41e4fd47" />

ravicious replied:
This one is rough. `defaultProps` on function components are no longer supported. This also affects components created through the `styled` helper. Class components are not affected.

I thought that components made with `styled` won't be affected since I didn't see warnings in the console about it.

https://github.com/gravitational/teleport/blob/96da21ab845d473206a0c60def551efd0b7e5009/web/packages/teleterm/src/ui/Vnet/DocumentVnetInfo.tsx#L265-L266
https://github.com/gravitational/teleport/blob/96da21ab845d473206a0c60def551efd0b7e5009/web/packages/teleterm/src/ui/Vnet/DocumentVnetInfo.tsx#L165-L184

ravicious replied:
It looks like we need to do something like this instead of using `defaultProps`.

```
styled.div.attrs(props => ({ foo: 'bar', ...props }))``
```

https://github.com/styled-components/styled-components/issues/3967#issuecomment-1480583099

In `TextPart` above, I specifically didn't put `flex` within `.attrs` so that it can be overridden.

ravicious committed:
Remove defaultProps

avatus committed:
pnpm dedupe

avatus commented:
looks like the ui group update caused the pnpm-lock conflict. ill check it out tomorrow

kimlisa committed:
revert test and fix stories

kimlisa commented on web/packages/shared/components/FieldCheckbox/FieldCheckbox.test.tsx:
```diff
@@ -52,7 +52,7 @@ test('controlled flow', async () => {
   let checkboxRef;
   function TestForm() {
-    const cbRefInternal = useRef();
+    const cbRefInternal = useRef(undefined);
```
should this be null? there's actually still a few more files that uses `undefined` but `null` for others but wasn't sure if there was a reason for it

ravicious replied:
This can be undefined, see https://github.com/gravitational/teleport/pull/54460#discussion_r2081237132.

kimlisa commented on web/packages/teleport/src/Console/DocumentSsh/Terminal/Terminal.tsx:
```diff
@@ -48,8 +48,8 @@ export interface TerminalProps {
 export const Terminal = forwardRef<TerminalRef, TerminalProps>((props, ref) => {
-  const termCtrlRef = useRef<XTermCtrl>();
-  const elementRef = useRef<HTMLDivElement>();
+  const termCtrlRef = useRef<XTermCtrl>(undefined);
+  const elementRef = useRef<HTMLDivElement>(null);
```
like here, i was curious why we used `undefined` for one and `null` for other?

ravicious replied:
Writing this up here so that we can refer later to it in the future. :~)

## Changes to `useRef` in React 19

In React 19, [`useRef` requires an argument](https://react.dev/blog/2024/04/25/react-19-upgrade-guide#useref-requires-argument). The type definition for it now looks like this:

```typescript
interface RefObject<T> {
  current: T
}

declare function useRef<T>: RefObject<T>
```

If you use strict null checks in TypeScript (like we do in the design package), this means that the ref object returned by `useRef(foo)` _always_ holds a value of type `Foo`.

However, sometimes you want what's in the ref object to be nullable. That's the case with [using refs to access DOM elements](https://react.dev/reference/react/useRef#manipulating-the-dom-with-a-ref). Once the node that used the ref is no longer in the DOM, React sets the ref object's value to null. So you essentially want to say something like this:

```tsx
const ref = useRef<HTMLDivElement | null>(null)
```

To make this easier to type because it's such a common pattern, there's an overload for `useRef` which makes `useRef<T>(null)` return `RefObject<T | null>`, so you can write this instead:

```tsx
const ref = useRef<HTMLDivElement>(null)
```

## `useRef(undefined)`

> To ease migration due to the required argument for `useRef`, a convenience overload for `useRef(undefined)` was added that automatically returns `RefObject<T | undefined>`.

This means that now every call site which previously used `useRef` without an argument was modified with a code mod to pass `undefined` as the argument instead. This matches the semantics from React 18: if you used `useRef` without an arg, `ref.value` was `undefined`.

## `null` vs `undefined`

* When dealing with refs passed to the DOM, always pass `null` since that's the value that React sets the ref value to when the ref gets removed from the DOM.
   * I believe that with strict null checks on, you might even get a type error if you try to use `undefined` here.
* When dealing with any other ref, it ultimately doesn't matter if you use `null` or `undefined`. `undefined` was added automatically through a code mod since technically that's the initial value used by callsites which called `useRef()` without an arg back in React 18.

kimlisa replied:
thanks @ravicious 🙏

kimlisa reviewed (approved):
:shipit:

ravicious committed:
Merge branch 'master' into avatus/react19

ravicious committed:
Import JSX in StatusInfo.tsx
